libs/core/algorithms/include/hpx/parallel/algorithms/nth_element.hpp

```diff
@@ -0,0 +1,411 @@
+    /// \a parallel_task_policy are permitted to execute in an unordered
+    /// fashion in unspecified threads, and indeterminately sequenced
+    /// within each thread.
+    ///
+    /// \returns  The \a nth_element algorithms returns nothing.
+    ///
+    template <typename ExPolicy, typename RandomIt, typename Pred>
+    void nth_element(ExPolicy&& policy, RandomIt first, RandomIt nth,
+        RandomIt last, Pred&& pred);
+
+    // clang-format on
+}    // namespace hpx
+
+#else    // DOXYGEN
+
+#include <hpx/assert.hpp>
+#include <hpx/config.hpp>
+#include <hpx/concepts/concepts.hpp>
+#include <hpx/functional/invoke.hpp>
+#include <hpx/iterator_support/traits/is_iterator.hpp>
+#include <hpx/parallel/util/detail/sender_util.hpp>
+
+#include <hpx/execution/algorithms/detail/predicates.hpp>
+#include <hpx/executors/execution_policy.hpp>
+#include <hpx/parallel/algorithms/detail/dispatch.hpp>
+#include <hpx/parallel/algorithms/minmax.hpp>
+#include <hpx/parallel/algorithms/partial_sort.hpp>
+#include <hpx/parallel/algorithms/partition.hpp>
+#include <hpx/parallel/util/detail/algorithm_result.hpp>
+
+#include <algorithm>
+#include <cstddef>
+#include <iterator>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace hpx { namespace parallel { inline namespace v1 {
+    ///////////////////////////////////////////////////////////////////////////
+    // nth_element
+    namespace detail {
+
+        ///////////////////////////////////////////////////////////////////////
+        ///
+        /// \brief : The element placed in the nth position is exactly the
+        ///          element that would occur in this position if the range
+        ///          was fully sorted. All of the elements before this new nth
+        ///          element are less than or equal to the elements after the
+        ///          new nth element.
+        ///
+        /// \param first : iterator to the first element
+        /// \param nth : iterator defining the sort partition point
+        /// \param end : iterator to the element after the last in the range
+        /// \param level : level of depth in the call from the root
+        /// \param comp : object for to Compare elements
+        /// \param proj : projection
+        ///
+        template <class RandomIt, typename Compare, typename Proj>
+        void nth_element_seq(RandomIt first, RandomIt nth, RandomIt end,
+            uint32_t level, Compare&& comp, Proj&& proj)
+        {
+            const uint32_t nmin_sort = 24;
+
+            // Check if the iterators are corrects
+            auto nelem = end - first;
+            HPX_ASSERT(nelem >= 0 and (nth - first + 1) > 0 and
+                (nth - first + 1) <= nelem);
+
+            // Check  the special conditions
+            if (nth == first)
+            {
+                RandomIt it = detail::min_element<RandomIt>().call(
+                    hpx::execution::seq, first, end, comp, proj);
```

Comment:
Should we `std::forward` the predicate and the projection here?

Reply:
Oh right yeah, line 193 too since we are returning immediately after there too.

Reply:
Yes, this is needed in several spots. I have mentioned it once only.